theano/misc/pkl_utils.py

```diff
@@ -93,3 +113,217 @@ class CompatUnpickler(pickle._Unpickler):
+
+    :param obj: The object to pickle.
+    :type obj: object
+
+    :param f: The file handle to save the object to.
+    :type f: file
+
+    :param protocol: The pickling protocol to use. Unlike Python's built-in
+        pickle, the default is set to `2` insstead of 0 for Python 2. The
+        Python 3 default (level 3) is maintained.
+    :type protocol: int, optional
+
+    :param persistent_id: The callable that persists certain objects in the
+        object hierarchy to separate files inside of the zip file. For example,
+        :class:`PersistentNdarrayID` saves any :class:`numpy.ndarray` to a
+        separate NPY file inside of the zip file.
+    :type persistent_id: callable
+
+
+    .. note::
+        The final file is simply a zipped file containing at least one file,
+        `pkl`, which contains the pickled object. It can contain any other
+        number of external objects. Note that the zip files are compatible with
+        NumPy's :func:`numpy.load` function.
+
+    >>> import theano
+    >>> foo_1 = theano.shared(0, name='foo')
+    >>> foo_2 = theano.shared(1, name='foo')
+    >>> with open('model.zip', 'w') as f:
+    ...     dump((foo_1, foo_2, numpy.array(2)), f)
+    >>> numpy.load('model.zip').keys()
+    ['foo', 'foo_2', 'array_0', 'pkl']
+    >>> numpy.load('model.zip')['foo']
+    array(0)
+    >>> with open('model.zip') as f:
+    ...     foo_1, foo_2, array = load(f)
+    >>> array
+    array(2)
+
+    """
+    with closing(zipfile.ZipFile(f, 'w', zipfile.ZIP_DEFLATED,
+                                 allowZip64=True)) as zip_file:
+        def func(f):
+            p = pickle.Pickler(f, protocol=protocol)
+            p.persistent_id = persistent_id(zip_file)
+            p.dump(obj)
+        zipadd(func, zip_file, 'pkl')
+
+
+def load(f, persistent_load=PersistentNdarrayLoad):
+    """Load a file that was dumped to a zip file.
+
+    :param f: The file handle to the zip file to load the object from.
+    :type f: file
+
+    :param persistent_load: The persistent loading function to use for
+        unpickling. This must be compatible with the `persisten_id` function
+        used when pickling.
+    :type persistent_load: callable, optional
+
+    """
+    with closing(zipfile.ZipFile(f, 'r')) as zip_file:
+        p = pickle.Unpickler(StringIO(zip_file.open('pkl').read()))
+        p.persistent_load = persistent_load(zip_file)
+        return p.load()
+
+
+def zipadd(func, zip_file, name):
+    """Calls a function with a file object, saving it to a zip file.
+
+    :param func: The function to call.
+    :type func: callable
+
+    :param zip_file: The zip file that `func` should write its data to.
+    :type zip_file: :class:`zipfile.ZipFile`
+
+    :param name: The name of the file inside of the zipped archive that `func`
+        should save its data to.
+    :type name: str
+
+    """
+    with tempfile.NamedTemporaryFile('wb', delete=False) as temp_file:
```

Comment:
It doesn't get deleted by Python, but it's in `/tmp` so I guess it will get deleted by the system eventually (i.e. on reboots). Might still be worth removing manually I guess.